cellarium/ml/transforms/transforms.py

```diff
@@ -1,108 +1,251 @@
+
+    .. math::
+
+        \\mathrm{mask}_g = \\mathrm{feature}_g \\in \\mathrm{filter\\_list}
+
+        y_{ng} = x_{ng}[:, \\mathrm{mask}_g]
+
+    Args:
+        filter_list: A list of features to filter by.
+    """
+
+    def __init__(self, filter_list: ArrayLike) -> None:
+        super().__init__()
+        self.filter_list = np.array(filter_list)
+
+    @cache
+    def filter(self, feature_g: tuple) -> np.ndarray[Any, np.dtype[np.bool_]]:
+        """
+        Args:
+            feature_g: The list of the variable names in the input data.
+
+        Returns:
+            A boolean mask of the features to filter by.
+        """
+        return np.isin(feature_g, self.filter_list)
+
+    def forward(self, x_ng: torch.Tensor, feature_g: np.ndarray) -> torch.Tensor:
+        """
+        Args:
+            x_ng:
+                Gene counts.
+            feature_g:
+                The list of the variable names in the input data.
+
+        Returns:
+            Filtered gene counts.
+        """
+        assert x_ng.shape[1] == len(feature_g), "The number of x_ng columns must match the feature_g length."
+
+        filter_mask = self.filter(tuple(feature_g.tolist()))
+        return x_ng[:, filter_mask]
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}(filter_list={self.filter_list})"
+
+
+class Log1p(nn.Module):
+    """
+    Log1p transform gene counts.
+
+    .. math::
+
+        y_{ng} = \\log(1 + x_{ng})
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
 
     def forward(self, x_ng: torch.Tensor) -> torch.Tensor:
-        return self.target_count * x_ng / (x_ng.sum(dim=-1, keepdim=True) + self.eps)
+        """
+        Args:
+            x_ng: Gene counts.
+
+        Returns:
+            Log1p transformed gene counts.
+        """
+        return torch.log1p(x_ng)
 
     def __repr__(self) -> str:
-        return f"{self.__class__.__name__}(target_count={self.target_count}, eps={self.eps})"
+        return f"{self.__class__.__name__}()"
 
 
-class DivideByScale(nn.Module):
+class NormalizeTotal(nn.Module):
     """
-    Divide gene counts by a scale.
+    Normalize total gene counts per cell to target count.
+
+    .. math::
+
+        \\mathrm{total\\_mrna\\_umis}_n = \\sum_{g=1}^G x_{ng}
+
+        y_{ng} = \\frac{\\mathrm{target\\_count} \\times x_{ng}}{\\mathrm{total\\_mrna\\_umis}_n + \\mathrm{eps}}
 
     Args:
-        scale_g:
-            A scale for each gene.
+        target_count:
+            Target gene epxression count.
         eps:
             A value added to the denominator for numerical stability.
     """
 
-    def __init__(self, scale_g: torch.Tensor, eps: float = 1e-6):
+    def __init__(
+        self,
+        target_count: int = 10_000,
+        eps: float = 1e-6,
+    ) -> None:
         super().__init__()
-        self.scale_g: torch.Tensor
-        self.register_buffer("scale_g", scale_g)
+        self.target_count = target_count
```

Comment:
assert that target_count is positive.

Reply:
write utility functions `assert_positive(Number) -> Number` and `assert_nonnegative(Number) -> Number`.